File: app/src/main/java/de/thomasstache/parklotse/State.java
package de.thomasstache.parklotse;

import android.content.SharedPreferences;

import com.mapbox.mapboxsdk.geometry.LatLng;

/**
 * The whole application state.
 */
public class State
{
	private static final String PREF_PARKED = "isParked";
	private static final String PREF_LOCATION_LON = "parkedLocationLongitude";
	private static final String PREF_LOCATION_LAT = "parkedLocationLatitude";

	public boolean isParked = false;

	// default location at Dresden Albertplatz
	public LatLng latLng = new LatLng(51.063, 13.746);

	public static State createFromPrefs(SharedPreferences preferences)
	{
		final State state = new State();

		state.isParked = preferences.getBoolean(PREF_PARKED, false);
		if (state.isParked)
		{
			// LatLng is only valid if we were parked
			String latitude = preferences.getString(PREF_LOCATION_LAT, "");
			String longitude = preferences.getString(PREF_LOCATION_LON, "");

			if (!latitude.isEmpty() && !longitude.isEmpty())
			{
				state.latLng = parseLatLng(latitude, longitude);
			}
		}

		return state;
	}

	public static boolean saveToPrefs(State state, SharedPreferences preferences)
	{
		final SharedPreferences.Editor editor = preferences.edit();

		final String latitude;
		final String longitude;

		if (state.latLng != null)
		{
			latitude = Double.toString(state.latLng.getLatitude());
			longitude = Double.toString(state.latLng.getLongitude());
		}
		else
		{
			latitude = "";
			longitude = "";
		}

		editor.putBoolean(PREF_PARKED, state.isParked)
		      .putString(PREF_LOCATION_LAT, latitude)
		      .putString(PREF_LOCATION_LON, longitude);

		return editor.commit();
	}

	private static LatLng parseLatLng(String latitude, String longitude)
	{
		final double lat = Double.parseDouble(latitude);
		final double lon = Double.parseDouble(longitude);

		// TODO add error checking

		return new LatLng(lat, lon);
	}
}

File: app/src/main/java/de/thomasstache/parklotse/State.kt
package de.thomasstache.parklotse

import android.content.SharedPreferences
import androidx.core.content.edit
import com.mapbox.mapboxsdk.geometry.LatLng
import java.lang.Double.parseDouble

/**
 * The whole application state.
 */
class State {
    /**
     * Indicates whether we are currently parked, and [latLng] has a valid value.
     */
    var isParked: Boolean = false

    // default location at Dresden Albertplatz
    var latLng: LatLng? = LatLng(51.063, 13.746)

    fun saveToPrefs(preferences: SharedPreferences) {
        val latitude: String = latLng?.latitude?.toString() ?: ""
        val longitude: String = latLng?.longitude?.toString() ?: ""

        preferences.edit {
            putBoolean(PREF_PARKED, isParked)
            putString(PREF_LOCATION_LAT, latitude)
            putString(PREF_LOCATION_LON, longitude)
        }
    }

    companion object {
        private const val PREF_PARKED = "isParked"
        private const val PREF_LOCATION_LON = "parkedLocationLongitude"
        private const val PREF_LOCATION_LAT = "parkedLocationLatitude"

        fun createFromPrefs(preferences: SharedPreferences): State {
            val state = State()

            state.isParked = preferences.getBoolean(PREF_PARKED, false)
            if (state.isParked) {
                // LatLng is only valid if we were parked
                val latitude = preferences.getString(PREF_LOCATION_LAT, "")!!
                val longitude = preferences.getString(PREF_LOCATION_LON, "")!!

                if (latitude.isNotEmpty() && longitude.isNotEmpty()) {
                    state.latLng = parseLatLng(latitude, longitude)
                }
            }

            return state
        }

        private fun parseLatLng(
            latitude: String,
            longitude: String,
        ): LatLng {
            val lat = parseDouble(latitude)
            val lon = parseDouble(longitude)

            // TODO add error checking

            return LatLng(lat, lon)
        }
    }
}
